api/src/main/java/org/pmiops/workbench/api/OAuth2ApiController.java

```diff
@@ -0,0 +1,85 @@
+package org.pmiops.workbench.api;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.servlet.http.HttpServletRequest;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.http.NameValuePair;
+import org.apache.http.client.utils.URLEncodedUtils;
+import org.apache.http.message.BasicNameValuePair;
+import org.pmiops.workbench.exceptions.BadRequestException;
+import org.pmiops.workbench.google.CloudStorageClient;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.http.client.HttpComponentsClientHttpRequestFactory;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.client.RestTemplate;
+
+@Controller
+public class OAuth2ApiController {
+  static final String AOU_TOKEN_ENDPOINT = "/oauth2/token";
+  static final String GOOGLE_TOKEN_ENDPOINT = "https://oauth2.googleapis.com/token";
+  private static final String CLIENT_SECRET_PARAM = "client_secret";
+  private final RestTemplate restTemplate;
+  private final CloudStorageClient cloudStorageClient;
+
+  @Autowired
+  public OAuth2ApiController(CloudStorageClient cloudStorageClient) {
+    restTemplate = new RestTemplate(new HttpComponentsClientHttpRequestFactory());
```

Comment:
> RestTemplate should be injected as a Bean

Done.

> what is the reason choosing RestTemplate with HttpComponentsClientHttpRequestFactory

I copied the code from Terra's implementation of this method.

I spent a couple hours trying to convert this to one of the other two options with no luck. Since I'm not making any progress I don't think it's worth additional effort to convert, especially since I'm unfamiliar with all three libraries. If you can convert it I'll swap out the code.

I haven't successfully created a GoogleHttpClient request at all. The public docs are very minimal as far as I can tell, and I'd prefer not to rely on Google-internal examples to learn how to use it.

For OkHttp, the main issue I'm having is that we ultimately need to return an instance of `org.springframework.http.ResponseEntity` instead of the OkHttp equivalent. `RestTemplate` returns the correct object type which makes the request seem like a true proxy.

Reply:
I agree with the overall intent of not doing the same thing many ways, though. It seems like we have far too many dependencies to pull from, generally, although I'm not sure if that's an AoU problem or a Java ecosystem problem.